### PR TITLE
[HF][build][test] add Huggingface Training compiler infra

### PR DESCRIPTION
**Pull request by Qingzi-Lan**
### Description
add Huggingface Training compiler infra
picked commit @fe8cbde

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: huggingface/pytorch/buildspec-trcomp.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    transformers_version: &TRANSFORMERS_VERSION 4.17.0
+    datasets_version: &DATASETS_VERSION 1.18.4
+    tag: !join [ *VERSION, '-', 'transformers', *TRANSFORMERS_VERSION, '-', *DEVICE_TYPE, '-', *TAG_PYTHON_VERSION, '-',
+                 *CUDA_VERSION, '-', *OS_VERSION ]
+    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /,
+                         *CUDA_VERSION, /Dockerfile.trcomp., *DEVICE_TYPE ]
+    download_artifacts:
+      pytorch_whl:
+        type: s3
+        URI: s3://sm-training-comp-pytorch-binaries/cd4b3932-8a23-4d21-95d7-c14e1351bef5/20220401-191238/4187a5ac613030093caa36c7d5ea95c3b9bb9b98/torch-1.10.2+cu113-cp38-cp38-linux_x86_64.whl
```
Comment (tejaschumbalkar) [inline huggingface/pytorch/buildspec-trcomp.yml:34]: we need to use the public wheel

File: huggingface/tensorflow/buildspec-trcomp.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    transformers_version: &TRANSFORMERS_VERSION 4.17.0
+    datasets_version: &DATASETS_VERSION 1.18.2
+    tag: !join [ *VERSION, '-', 'transformers', *TRANSFORMERS_VERSION, '-', *DEVICE_TYPE, '-', *TAG_PYTHON_VERSION, '-',
+                 *CUDA_VERSION, '-', *OS_VERSION ]
+    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /,
+                         *CUDA_VERSION, /Dockerfile.trcomp., *DEVICE_TYPE ]
+    download_artifacts:
+      tensorflow_whl:
+        type: s3
+        URI: s3://framework-binaries/tensorflow/trcomp/r2.6_aws/gpu/2022-03-20-05-33/tensorflow_gpu-2.6.3-cp38-cp38-manylinux_2_12_x86_64.manylinux2010_x86_64.whl
```
Comment (tejaschumbalkar) [inline huggingface/tensorflow/buildspec-trcomp.yml:34]: same here, a public wheel is required and should be directly installed in the dockerfile. The configuration in the buildspec file should be removed.

File: src/image_builder.py [hunk 3]
```diff
@@ -50,6 +51,21 @@ def _find_image_object(images_list, image_name):
     return ret_image_object
 
 
+def is_private_artifact_label_required(artifact_uri, image_repo_uri):
```
Comment (tejaschumbalkar) [inline src/image_builder.py:54]: if this function is only required for private repo, lets not add it in public repo

File: test/dlc_tests/conftest.py [hunk 4]
```diff
@@ -711,18 +720,24 @@ def lookup_condition(lookup, image):
...
+    #Extract ecr repo name from the image and check if it exactly matches the lookup (fixture name)
+    repo_name = image.split("/")[-1].split(":")[0]
 
-    job_types = (
-        "training",
-        "inference",
-    )
-    device_types = ("cpu", "gpu", "eia", "neuron", "hpu", "graviton")
+    job_type = ("training", "inference",)
+    device_type = ("cpu", "gpu", "eia", "neuron", "hpu")
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:727]: add `graviton` to the list

File: test/test_utils/__init__.py [hunk 5]
```diff
@@ -1265,6 +1289,21 @@ def get_synapseai_version_from_tag(image_uri):
     return synapseai_version
 
 
+def get_synapseai_version_from_tag(image_uri):
```
Comment (tejaschumbalkar) [inline test/test_utils/__init__.py:1292]: this is a duplicate function. we should remove this from private repo __init__.py file

Comment (Qingzi-Lan) [reply test/test_utils/__init__.py:1292]: remove from PR, I see only one `get_synapseai_version_from_tag` in private repo [here](https://github.com/aws/PRIVATE-deep-learning-containers/blob/private-master/test/test_utils/__init__.py#L1239).

File: test/dlc_tests/conftest.py [hunk 6]
```diff
@@ -711,18 +720,24 @@ def lookup_condition(lookup, image):
     """
     Return true if the ECR repo name ends with the lookup or lookup contains job type or device type part of the image uri.
     """
-    # Extract ecr repo name from the image and check if it exactly matches the lookup (fixture name)
-    repo_name = get_ecr_repo_name(image)
+    #Extract ecr repo name from the image and check if it exactly matches the lookup (fixture name)
+    repo_name = image.split("/")[-1].split(":")[0]
```
Comment (tejaschumbalkar) [inline test/dlc_tests/conftest.py:724]: why is this replaced by `get_ecr_repo_name`?

Comment (Qingzi-Lan) [reply test/dlc_tests/conftest.py:724]: it should maintain the same using `get_ecr_repo_name`

**Review by tejaschumbalkar — DISMISSED** [review]

**Review by Lokiiiiii — APPROVED** [review]

**Review by tejaschumbalkar — APPROVED** [review]

**Pull request merged**